`CommonSpaces` are not compatible with MPI

To reproduce, run the `unit_common_spaces.jl` test to initialize the MPI context. To do so, just add 
```julia
ClimaComms.init(ClimaComms.context())
```
before the testset. 

Run the code with mpi, the result is:
```julia
  AssertionError: context isa ClimaComms.SingletonCommsContext
  Stacktrace:
    [1] _IntervalTopology(context::ClimaComms.MPICommsContext{ClimaComms.CPUSingleThreaded, MPI.Comm}, mesh::ClimaCore.Meshes.IntervalMesh{ClimaCore.Meshes.Uniform, ClimaCore.Domains.IntervalDomain{ClimaCore.Geometry.ZPoint{Float64}, Tuple{Symbol, Symbol}}, LinRange{ClimaCore.Geometry.ZPoint{Float64}, Int64}, Nothing})
      @ ClimaCore.Topologies ~/repos/ClimaCore.jl/src/Topologies/interval.jl:46
    [2] (::ClimaCore.Topologies.var"#1#2"{ClimaComms.MPICommsContext{ClimaComms.CPUSingleThreaded, MPI.Comm}, ClimaCore.Meshes.IntervalMesh{ClimaCore.Meshes.Uniform, ClimaCore.Domains.IntervalDomain{ClimaCore.Geometry.ZPoint{Float64}, Tuple{Symbol, Symbol}}, LinRange{ClimaCore.Geometry.ZPoint{Float64}, Int64}, Nothing}})()
      @ ClimaCore.Topologies ~/repos/ClimaCore.jl/src/Topologies/interval.jl:37
    [3] get!
      @ ./dict.jl:458 [inlined]
    [4] IntervalTopology
      @ ~/repos/ClimaCore.jl/src/Topologies/interval.jl:36 [inlined]
    [5] ExtrudedCubedSphereGrid(::Type{Float64}; z_elem::Int64, z_min::Int64, z_max::Int64, radius::Int64, h_elem::Int64, n_quad_points::Int64, device::ClimaComms.CPUSingleThreaded, context::ClimaComms.MPICommsContext{ClimaComms.CPUSingleThreaded, MPI.Comm}, stretch::ClimaCore.Meshes.Uniform, hypsography_fun::ClimaCore.CommonGrids.var"#4#6", global_geometry::ClimaCore.Geometry.ShallowSphericalGlobalGeometry{Float64}, quad::ClimaCore.Quadratures.GLL{4}, h_mesh::ClimaCore.Meshes.EquiangularCubedSphere{ClimaCore.Domains.SphereDomain{Float64}, ClimaCore.Meshes.NormalizedBilinearMap}, h_topology::ClimaCore.Topologies.Topology2D{ClimaComms.MPICommsContext{ClimaComms.CPUSingleThreaded, MPI.Comm}, ClimaCore.Meshes.EquiangularCubedSphere{ClimaCore.Domains.SphereDomain{Float64}, ClimaCore.Meshes.NormalizedBilinearMap}, CartesianIndices{3, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}, Base.OneTo{Int64}}}, LinearIndices{3, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}, Base.OneTo{Int64}}}, Vector{Tuple{Int64, Int64, Int64, Int64, Bool}}, Vector{Tuple{Int64, Int64, Int64, Int64, Bool}}, Vector{Tuple{Int64, Int64}}, Vector{Int64}, Vector{Tuple{Bool, Int64, Int64}}, Vector{Int64}, Vector{Int64}, @NamedTuple{}, Vector{Tuple{Int64, Int64}}}, horizontal_layout_type::Type, z_mesh::ClimaCore.Meshes.IntervalMesh{ClimaCore.Meshes.Uniform, ClimaCore.Domains.IntervalDomain{ClimaCore.Geometry.ZPoint{Float64}, Tuple{Symbol, Symbol}}, LinRange{ClimaCore.Geometry.ZPoint{Float64}, Int64}, Nothing}, enable_bubble::Bool)
      @ ClimaCore.CommonGrids ~/repos/ClimaCore.jl/src/CommonGrids/CommonGrids.jl:192
    [6] ExtrudedCubedSphereSpace(::Type{Float64}; staggering::CellCenter, kwargs::@Kwargs{z_elem::Int64, z_min::Int64, z_max::Int64, radius::Int64, h_elem::Int64, n_quad_points::Int64, horizontal_layout_type::UnionAll})
      @ ClimaCore.CommonSpaces ~/repos/ClimaCore.jl/src/CommonSpaces/CommonSpaces.jl:121
    [7] ExtrudedCubedSphereSpace
      @ ~/repos/ClimaCore.jl/src/CommonSpaces/CommonSpaces.jl:121 [inlined]
    [8] #ExtrudedCubedSphereSpace#1
      @ ~/repos/ClimaCore.jl/src/CommonSpaces/CommonSpaces.jl:119 [inlined]
    [9] macro expansion
      @ ~/repos/ClimaCore.jl/test/CommonSpaces/unit_common_spaces.jl:32 [inlined]
   [10] macro expansion
      @ ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/Test/src/Test.jl:1704 [inlined]
   [11] top-level scope
      @ ~/repos/ClimaCore.jl/test/CommonSpaces/unit_common_spaces.jl:23
   [12] include(mod::Module, _path::String)
      @ Base ./Base.jl:557
   [13] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:323
   [14] _start()
```

Something is not initialized with the correct context

--- a/.buildkite/pipeline.yml
+++ b/.buildkite/pipeline.yml
@@ -273,6 +273,35 @@ steps:
         agents:
           slurm_gpus: 1
 
+      - label: "Unit: common spaces with CUDA"
+        key: "gpu_common_cuda_spaces"
+        command:
+          - "julia --color=yes --check-bounds=yes --project=.buildkite test/CommonSpaces/unit_common_spaces.jl"
+        env:
+          CLIMACOMMS_DEVICE: "CUDA"
+        agents:
+          slurm_gpus: 1
+
+      - label: "Unit: common spaces with CUDA and MPI"
+        key: "gpu_common_cuda_mpi_spaces"
+        command:
+          - "srun julia --color=yes --check-bounds=yes --project=.buildkite test/CommonSpaces/unit_common_spaces.jl"
+        env:
+          CLIMACOMMS_CONTEXT: "MPI"
+          CLIMACOMMS_DEVICE: "CUDA"
+        agents:
+          slurm_gpus: 1
+          slurm_ntasks: 2
+
+      - label: "Unit: common spaces with MPI"
+        key: "common_cuda_mpi_spaces"
+        command:
+          - "srun julia --color=yes --check-bounds=yes --project=.buildkite test/CommonSpaces/unit_common_spaces.jl"
+        env:
+          CLIMACOMMS_CONTEXT: "MPI"
+        agents:
+          slurm_ntasks: 2
+
       - label: "Unit: distributed cuda spaces"
         key: "gpu_distributed_extruded_cuda_spaces"
         command:

--- a/NEWS.md
+++ b/NEWS.md
@@ -4,6 +4,14 @@ ClimaCore.jl Release Notes
 main
 -------
 
+- Prior to this version, `CommonSpaces` could not be created with
+`ClimaComms.MPICommContext`. This is now fixed with PR
+[2176](https://github.com/CliMA/ClimaCore.jl/pull/2176).
+
+
+v0.14.24
+-------
+
  - A new `Adapt` wrapper was added, `to_device`, which allows users to adapt datalayouts, spaces, fields, and fieldvectors between the cpu and gpu. PR [2159](https://github.com/CliMA/ClimaCore.jl/pull/2159).
  - Remap interpolation cuda threading was improved. PR [2159](https://github.com/CliMA/ClimaCore.jl/pull/2159).
  - `center_space` and `face_space` are now exported from `CommonSpaces`. PR [2157](https://github.com/CliMA/ClimaCore.jl/pull/2157).

--- a/src/CommonGrids/CommonGrids.jl
+++ b/src/CommonGrids/CommonGrids.jl
@@ -189,7 +189,10 @@ function ExtrudedCubedSphereGrid(
         horizontal_layout_type,
         enable_bubble,
     )
-    z_topology = Topologies.IntervalTopology(context, z_mesh)
+    z_topology = Topologies.IntervalTopology(
+        ClimaComms.SingletonCommsContext(device),
+        z_mesh,
+    )
     z_grid = Grids.FiniteDifferenceGrid(z_topology)
     return Grids.ExtrudedFiniteDifferenceGrid(
         h_grid,
@@ -307,6 +310,7 @@ function ColumnGrid(
     ),
 ) where {FT}
     @assert ClimaComms.device(context) == device "The given device and context device do not match."
+    @assert context isa ClimaComms.SingletonCommsContext "Columns can only be created on Singleton contextes."
     z_topology = Topologies.IntervalTopology(context, z_mesh)
     return Grids.FiniteDifferenceGrid(z_topology)
 end
@@ -439,7 +443,10 @@ function Box3DGrid(
         horizontal_layout_type,
         enable_bubble,
     )
-    z_topology = Topologies.IntervalTopology(context, z_mesh)
+    z_topology = Topologies.IntervalTopology(
+        ClimaComms.SingletonCommsContext(device),
+        z_mesh,
+    )
     z_grid = Grids.FiniteDifferenceGrid(z_topology)
     return Grids.ExtrudedFiniteDifferenceGrid(
         h_grid,
@@ -542,10 +549,16 @@ function SliceXZGrid(
     @assert horizontal_layout_type <: DataLayouts.AbstractData
     @assert ClimaComms.device(context) == device "The given device and context device do not match."
 
-    h_topology = Topologies.IntervalTopology(context, h_mesh)
+    h_topology = Topologies.IntervalTopology(
+        ClimaComms.SingletonCommsContext(device),
+        h_mesh,
+    )
     h_grid =
         Grids.SpectralElementGrid1D(h_topology, quad; horizontal_layout_type)
-    z_topology = Topologies.IntervalTopology(context, z_mesh)
+    z_topology = Topologies.IntervalTopology(
+        ClimaComms.SingletonCommsContext(device),
+        z_mesh,
+    )
     z_grid = Grids.FiniteDifferenceGrid(z_topology)
     return Grids.ExtrudedFiniteDifferenceGrid(
         h_grid,

--- a/test/CommonSpaces/unit_common_spaces.jl
+++ b/test/CommonSpaces/unit_common_spaces.jl
@@ -16,6 +16,9 @@ using ClimaCore:
     DataLayouts
 using Test
 
+# Initialize MPI context
+ClimaComms.init(ClimaComms.context())
+
 @testset "Convenience constructors" begin
     function warp_surface(coord)
         # sin²(x) form ground elevation
@@ -75,14 +78,17 @@ using Test
     @test grid isa Grids.SpectralElementGrid2D
     @test Grids.topology(grid).mesh isa Meshes.EquiangularCubedSphere
 
-    space = ColumnSpace(;
-        z_elem = 10,
-        z_min = 0,
-        z_max = 1,
-        staggering = Grids.CellCenter(),
-    )
-    grid = Spaces.grid(space)
-    @test grid isa Grids.FiniteDifferenceGrid
+    # Column spaces are not supported with MPI
+    if !(ClimaComms.context() isa ClimaComms.MPICommsContext)
+        space = ColumnSpace(;
+            z_elem = 10,
+            z_min = 0,
+            z_max = 1,
+            staggering = Grids.CellCenter(),
+        )
+        grid = Spaces.grid(space)
+        @test grid isa Grids.FiniteDifferenceGrid
+    end
 
     space = Box3DSpace(;
         z_elem = 10,
@@ -104,21 +110,24 @@ using Test
     @test grid.horizontal_grid isa Grids.SpectralElementGrid2D
     @test Grids.topology(grid.horizontal_grid).mesh isa Meshes.RectilinearMesh
 
-    space = SliceXZSpace(;
-        z_elem = 10,
-        x_min = 0,
-        x_max = 1,
-        z_min = 0,
-        z_max = 1,
-        periodic_x = false,
-        n_quad_points = 4,
-        x_elem = 4,
-        staggering = Grids.CellCenter(),
-    )
-    grid = Spaces.grid(space)
-    @test grid isa Grids.ExtrudedFiniteDifferenceGrid
-    @test grid.horizontal_grid isa Grids.SpectralElementGrid1D
-    @test Grids.topology(grid.horizontal_grid).mesh isa Meshes.IntervalMesh
+    # Slices are currently not compatible with GPU
+    if !(ClimaComms.device() isa ClimaComms.CUDADevice)
+        space = SliceXZSpace(;
+            z_elem = 10,
+            x_min = 0,
+            x_max = 1,
+            z_min = 0,
+            z_max = 1,
+            periodic_x = false,
+            n_quad_points = 4,
+            x_elem = 4,
+            staggering = Grids.CellCenter(),
+        )
+        grid = Spaces.grid(space)
+        @test grid isa Grids.ExtrudedFiniteDifferenceGrid
+        @test grid.horizontal_grid isa Grids.SpectralElementGrid1D
+        @test Grids.topology(grid.horizontal_grid).mesh isa Meshes.IntervalMesh
+    end
 
     space = RectangleXYSpace(;
         x_min = 0,

--- a/test/CommonSpaces/unit_common_spaces.jl
+++ b/test/CommonSpaces/unit_common_spaces.jl
@@ -37,7 +37,7 @@ ClimaComms.init(ClimaComms.context())
         h_elem = 10,
         n_quad_points = 4,
         horizontal_layout_type = DataLayouts.IJHF,
-        staggering = Grids.CellCenter(),
+        staggering = CellCenter(),
     )
     grid = Spaces.grid(space)
     @test grid isa Grids.ExtrudedFiniteDifferenceGrid
@@ -63,7 +63,7 @@ ClimaComms.init(ClimaComms.context())
         h_elem = 10,
         n_quad_points = 4,
         hypsography_fun,
-        staggering = Grids.CellCenter(),
+        staggering = CellCenter(),
     )
     grid = Spaces.grid(space)
     @test grid isa Grids.ExtrudedFiniteDifferenceGrid
@@ -103,7 +103,7 @@ ClimaComms.init(ClimaComms.context())
         n_quad_points = 4,
         x_elem = 3,
         y_elem = 4,
-        staggering = Grids.CellCenter(),
+        staggering = CellCenter(),
     )
     grid = Spaces.grid(space)
     @test grid isa Grids.ExtrudedFiniteDifferenceGrid
@@ -121,7 +121,7 @@ ClimaComms.init(ClimaComms.context())
             periodic_x = false,
             n_quad_points = 4,
             x_elem = 4,
-            staggering = Grids.CellCenter(),
+            staggering = CellCenter(),
         )
         grid = Spaces.grid(space)
         @test grid isa Grids.ExtrudedFiniteDifferenceGrid

